stats/out_graphite.go

```diff
@@ -0,0 +1,118 @@
+package stats
+
+import (
+	"bytes"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/raintank/worldping-api/pkg/log"
+)
+
+type GraphiteMetric interface {
+	// Report the measurements in graphite format and reset measurements for the next interval if needed
+	ReportGraphite(prefix []byte, buf []byte, now time.Time) []byte
+}
+
+type Graphite struct {
+	prefix                   []byte
+	addr                     string
+	prefixCounter32          string
+	prefixGauge32            string
+	prefixGauge64            string
+	prefixRange32            string
+	prefixLatencyHistogram32 string
+
+	toGraphite chan []byte
+}
+
+func NewGraphite(prefix, addr string, interval int, bufferSize int) {
+	if len(prefix) != 0 && prefix[len(prefix)-1] != '.' {
+		prefix = prefix + "."
+	}
+	g := &Graphite{
+		prefix:     []byte(prefix),
+		addr:       addr,
+		toGraphite: make(chan []byte, bufferSize),
+	}
+	go g.writer()
+	go g.reporter(interval)
+}
+
+func (g *Graphite) reporter(interval int) {
+	ticker := tick(time.Duration(interval) * time.Second)
+	for now := range ticker {
+		log.Debug("stats flushing for", now, "to graphite")
+		if cap(g.toGraphite) != 0 && len(g.toGraphite) == cap(g.toGraphite) {
+			// no space in buffer, no use in doing any work
+			continue
+		}
+
+		buf := make([]byte, 0)
+
+		var fullPrefix bytes.Buffer
+		for name, metric := range registry.list() {
+			fullPrefix.Reset()
+			fullPrefix.Write(g.prefix)
+			fullPrefix.WriteString(name)
+			fullPrefix.WriteRune('.')
+			buf = metric.ReportGraphite(fullPrefix.Bytes(), buf, now)
+		}
+		g.toGraphite <- buf
+	}
+}
+
+// graphiteWriter is the background workers that connects to graphite and submits all pending data to it
+// TODO: conn.Write() returns no error for a while when the remote endpoint is down, the reconnect happens with a delay
+func (g *Graphite) writer() {
+	connectTicker := time.Tick(time.Second)
+
+	lock := &sync.Mutex{}
+	var conn net.Conn
+
+	var err error
+	go func() {
+		for range connectTicker {
```

Comment:
Actually now that I think about it I'm wondering where there need to be two go routines, one to reconnect and one to write the data into the connection. It seems like that could easily be done with only one go routine and without the reconnect ticker if the reconnect gets triggered as soon as a write fails, or am I overlooking something?